tests/raftstore/test_snap.rs

```diff
@@ -194,34 +194,40 @@ fn test_server_snap_gc() {
     test_snap_gc(&mut cluster);
 }
 
+/// A helper function for testing the handling of snapshot is correct
+/// when there are multiple snapshots which have overlapped region ranges
+/// arrive at the same raftstore.
 fn test_concurrent_snap<T: Simulator>(cluster: &mut Cluster<T>) {
-    // util::init_log();
-    // disable raft log gc.
+    // Disable raft log gc in this test case.
     cluster.cfg.raft_store.raft_log_gc_tick_interval = 60000;
 
     let pd_client = cluster.pd_client.clone();
     // Disable default max peer count check.
     pd_client.disable_default_rule();
+
     let r1 = cluster.run_conf_change();
     cluster.must_put(b"k1", b"v1");
     pd_client.must_add_peer(r1, new_peer(2, 2));
-    // peer 2 can't step to leader.
+    // Force peer 2 to be followers all the way.
     cluster.add_send_filter(CloneFilterFactory(RegionPacketFilter::new(r1, 2)
         .msg_type(MessageType::MsgRequestVote)
         .direction(Direction::Send)));
     cluster.must_transfer_leader(r1, new_peer(1, 1));
     cluster.must_put(b"k3", b"v3");
-    // Now peer 1 can't send snapshot to peer 2
-    cluster.sim.wl().add_recv_filter(3, box PauseFirstSnapshotFilter::new());
-    cluster.sim.wl().add_recv_filter(1, box PauseFirstSnapshotFilter::new());
+    // Pile up snapshots of overlapped region ranges and deliver them all at once.
+    let (tx, rx) = mpsc::channel();
+    cluster.sim.wl().add_recv_filter(3, box CollectSnapshotFilter::new(tx));
     pd_client.must_add_peer(r1, new_peer(3, 3));
     let region = cluster.get_region(b"k1");
-    // TODO: should check whether snapshot has been sent before split.
+    // Ensure the snapshot of range ("", "") is sent and piled in filter.
+    if let Err(e) = rx.recv_timeout(Duration::from_secs(1)) {
+        panic!("the snapshot is not sent before split, e: {:?}", e);
+    }
+    // Split the region range and then there should be another snapshot for the split ranges.
     cluster.must_split(&region, b"k2");
     must_get_equal(&cluster.get_engine(3), b"k3", b"v3");
     cluster.must_put(b"k4", b"v4");
     must_get_equal(&cluster.get_engine(3), b"k4", b"v4");
```

Comment:
The old test is not strong enough.

This case is to check whether two overlap snapshots have been handled in a round of handle ready. If they do, `region_ranges` will have two overlap ranges. This case make two snapshot arrive at a time, but not check whether `region_ranges` has two overlap ranges. So we should put a new pair of kv, `("k11", "v11")` for example, to origin region, and assert that the new kv should exist in store 3. If it does, split command must be executed, if there are two overlap regions in `region_ranges`, split command should have panic, hence test should fail.